### PR TITLE
COR-450 add validation for recipient forms

### PR DESCRIPTION
**Pull request by ludydoo**
Added validation for "Recipient" forms

The validation happens in two parts
1. The static validation, things we can infer without having to make any database calls. That lies in the default "validation" package, where all other validation currently lies.
2. The dynamic validation, that happens for now in the POST /formdefinitions handler. Here, we must make a DB call to check if the "parent" of a recipient form is also a recipient form.

This PR is based off COR-499 branch, so I set the PR target to the COR-499 branch until the other one gets merged into main

### CODE REVIEW

File: pkg/api/types/validation/recipient_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+				FieldType: types.FieldType{
+					Text: &types.FieldTypeText{},
+				},
+			},
+			wantFieldIndex: 0,
+			wantErrs:       nil,
+		},
+		{
+			name: "with multiple key field",
+			fields: []*types.FieldDefinition{
```
Comment (internetti) [inline pkg/api/types/validation/recipient_test.go:50]: could we also add a test for testing if that is valid for non-recipients? I don't see where this example form is defined as a recipient form, so this might as well be valid

Comment (ludydoo) [reply pkg/api/types/validation/recipient_test.go:50]: I`m not sure how to write this test.. 
But this validation is skipped for non-recipient forms. Can you clarify what you would change ?

```
// ValidateRecipientForm validates recipient forms
func ValidateRecipientForm(form *types.FormDefinition) validation.ErrorList {
	var result validation.ErrorList
	if form.Type != types.RecipientFormType {
		return result
	}
         ...
```

Comment (internetti) [reply pkg/api/types/validation/recipient_test.go:50]: okay, true, that doesn't make sense then. but we could add a test that ValidateRecipientForm really skips the validation/returns an empty error list when the form is not a recipient form?

File: pkg/server/formsapi/handlers/form/create.go [hunk 2]
```diff
@@ -56,6 +62,47 @@ func (h *Handler) Create() http.HandlerFunc {
 	}
 }
 
+func validateRecipientFormParent(
```
Comment (internetti) [inline pkg/server/formsapi/handlers/form/create.go:65]: shouldn't this also live in validation/recipient.go?

Comment (internetti) [reply pkg/server/formsapi/handlers/form/create.go:65]: @ludydoo

File: pkg/api/types/validation/recipient_test.go (hunk 1 above)
Comment (ludydoo) [reply pkg/api/types/validation/recipient_test.go:50]: Sure, will do !

File: pkg/server/formsapi/handlers/form/create.go (hunk 2 above)
Comment (ludydoo) [reply pkg/server/formsapi/handlers/form/create.go:65]: @internetti the problem with this one is that it requires an access to the database. All other validation happening in validation/recipient.go is "static" (does not need any service call, etc.). Do you think it would be cleaner if it would be in the validation/recipient.go ?

Comment (ludydoo) [reply pkg/server/formsapi/handlers/form/create.go:65]: This piece of code should probably live in some kind of controller/validation webhook I think. But we don't have that layer yet !

Comment (neb42) [reply pkg/server/formsapi/handlers/form/create.go:65]: What about passing the store to `ValidateForm`?

**Comment by neb42** [comment]
lgtm, but @internetti I'll let you finish before approving

File: pkg/api/types/validation/recipient_test.go (hunk 1 above)
Comment (internetti) [reply pkg/api/types/validation/recipient_test.go:88]: @ludydoo it's a detail, but in the test you added, you test ValidateForm instead of ValidateRecipientForm, is there a reason for that?

File: pkg/server/formsapi/handlers/form/create.go (hunk 2 above)
Comment (ludydoo) [reply pkg/server/formsapi/handlers/form/create.go:65]: @neb42 that would work, though it would require all tests to inject the store as well. Imho I really think this should be on a controller layer of some kind.

File: pkg/api/types/validation/recipient_test.go (hunk 1 above)
Comment (ludydoo) [reply pkg/api/types/validation/recipient_test.go:88]: @internetti I wanted the test to also validate that the ValidateRecipientForm was called by the main "ValidateForm"

**Comment by neb42** [comment]
@ludydoo I don't think this is handling the case where there are no key fields.

**Comment by neb42** [comment]
I guess this is because of root forms where that is the correct validation. I feel validating a recipient form and a root recipient form are different things, and the root flow shouldn't be exposed to the user.

**Comment by ludydoo** [comment]
@neb42 I thought we agreed to handle this using access control

**Comment by neb42** [comment]
Yeah that rings a bell. Something like:
```
if "super-user" not in userRoles:
    validateRecipientHasKey()
```
